.github/workflows/docker.yml

```diff
@@ -0,0 +1,63 @@
+name: Build & Push Docker Images
+
+on:
+  workflow_dispatch:
+  pull_request:
+    branches:
+    - develop
```

Comment:
We should not run this on PRs as secrets are not passed down to PRs for security reasons. This should only run on a push to develop branch.

Reply:
> Yes I was just running it on this PR in order to test, this workflow can be triggered manually or run periodically afterwards, whichever preferred.

Reply:
Should be ready to merge once `pull_request` is removed. Now that I think of it, we should also add a cron job that builds and pushes docker images nightly? Pushing on every commit might be a bit excessive.

Reply:
I think we should also check out other repositories and projects as to how they do it (most related workflows push to the development branch and overwrite the previous image). Upon multiple pushes to `develop` we can cancel intermediate builds here as we do for the unit tests and benchmarks workflows

Reply:
> I think we should also check out other repositories and projects as to how they do it

+1!

Reply:
The first workflow I looked at was `PyHF`'s, of course. They build docker images on PRs to their development branch (`main` in their case), pushes to `main`, scheduled nightly, on version tags, for releases, and also have a workflow dispatch trigger.

Building and pushing images on release tags _might_ require configuring the naming scheme of the images to include the version since we would not want to overwrite any existing images. We should look into that before `23.9` comes out (additionally `23.9rcX`). For now, I think just the two options: pushes to `develop` and a workflow dispatch trigger should be fine, with concurrency checks to cancel the build in case two PRs are merged one after the other. The primary disadvantage of the nightlies is that it could cause some hinderances for developers working within their containers in case a large PR or one with breaking changes gets merged and the Dockerfile is not updated, though this would be a rare occurrence—developers have to deal with merge conflicts sometimes anyway, and this is the case without Docker as well.

It is possible to exercise more fine-grained control over how we can push to Docker Hub. Some data science projects use a workflow call trigger with particular event webhooks to run a workflow from other workflows. To say it is complicated is an understatement but this way, in theory, one can build the Docker images only at the completion of a particular workflow by calling another workflow through it. In this case, it would be when the tests pass on merging a PR to `develop`, which, in-turn, will be cancelled on subsequent pushes due to the concurrency setting (or do we not run CI on merge commits anymore?) We should explore this option only if the solution above seems to be excessive or just switch to the nightlies.

Reply:
`push` and `workflow_dispatch` sounds good to me as well!